Python/Product/PythonTools/PythonTools/Navigation/Navigable/NavigableSymbolSource.cs

```diff
@@ -0,0 +1,103 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Intellisense;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Language.StandardClassification;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Operations;
+
+namespace Microsoft.PythonTools.Navigation.Navigable {
+    class NavigableSymbolSource : INavigableSymbolSource {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _buffer;
+        private readonly IClassifier _classifier;
+        private readonly ITextStructureNavigator _textNavigator;
+        private readonly AnalysisEntryService _entryService;
+
+        private static readonly string[] _classifications = new string[] {
+            PredefinedClassificationTypeNames.Identifier,
+            PythonPredefinedClassificationTypeNames.Class,
+            PythonPredefinedClassificationTypeNames.Function,
+            PythonPredefinedClassificationTypeNames.Module,
+            PythonPredefinedClassificationTypeNames.Parameter,
+        };
+
+        public NavigableSymbolSource(IServiceProvider serviceProvider, ITextView textView, ITextBuffer buffer, IClassifier classifier, ITextStructureNavigator textNavigator) {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _textView = textView ?? throw new ArgumentNullException(nameof(textView));
+            _buffer = buffer ?? throw new ArgumentNullException(nameof(buffer));
+            _classifier = classifier ?? throw new ArgumentNullException(nameof(classifier));
+            _textNavigator = textNavigator ?? throw new ArgumentNullException(nameof(textNavigator));
+            _entryService = serviceProvider.GetEntryService();
+        }
+
+        public void Dispose() {
+        }
+
+        public async Task<INavigableSymbol> GetNavigableSymbolAsync(SnapshotSpan triggerSpan, CancellationToken cancellationToken) {
+            Debug.Assert(triggerSpan.Length == 1);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var extent = _textNavigator.GetExtentOfWord(triggerSpan.Start);
+            if (!extent.IsSignificant) {
+                return null;
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            AnalysisEntry entry = null;
+            _entryService?.TryGetAnalysisEntry(_textView, _buffer, out entry);
+            if (entry == null) {
+                return null;
+            }
+
+            foreach (var token in _classifier.GetClassificationSpans(extent.Span)) {
```

Comment:
Split this part down into a separate `internal` function - `static` if possible - and we can test that.